core/trino-main/src/test/java/io/trino/sql/TestExpressionInterpreter.java

```diff
@@ -638,6 +638,74 @@ public void testCastBigintToBoundedVarchar()
                 .hasMessage("Value -12300000000 cannot be represented as varchar(3)");
     }
 
+    @Test
+    public void testCastIntegerToBoundedVarchar()
+    {
+        assertEvaluatedEquals("CAST(1234 AS varchar(4))", "'1234'");
+        assertEvaluatedEquals("CAST(1234 AS varchar(50))", "'1234'");
+
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(1234 AS varchar(3))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value 1234 cannot be represented as varchar(3)");
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(-1234 AS varchar(3))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value -1234 cannot be represented as varchar(3)");
+    }
+
+    @Test
+    public void testCastSmallintToBoundedVarchar()
+    {
+        assertEvaluatedEquals("CAST(SMALLINT '1234' AS varchar(4))", "'1234'");
+        assertEvaluatedEquals("CAST(SMALLINT '1234' AS varchar(50))", "'1234'");
+
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(SMALLINT '1234' AS varchar(3))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value 1234 cannot be represented as varchar(3)");
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(SMALLINT '-1234' AS varchar(3))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value -1234 cannot be represented as varchar(3)");
+    }
+
+    @Test
+    public void testCastTinyintToBoundedVarchar()
+    {
+        assertEvaluatedEquals("CAST(TINYINT '123' AS varchar(3))", "'123'");
+        assertEvaluatedEquals("CAST(TINYINT '123' AS varchar(50))", "'123'");
+
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(TINYINT '123' AS varchar(2))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value 123 cannot be represented as varchar(2)");
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(TINYINT '-123' AS varchar(2))"))
+                .hasErrorCode(INVALID_CAST_ARGUMENT)
+                .hasMessage("Value -123 cannot be represented as varchar(2)");
+    }
+
+    @Test
+    public void testCastDecimalToBoundedVarchar()
+    {
+        // short decimal
+        assertEvaluatedEquals("CAST(DECIMAL '12.4' AS varchar(4))", "'12.4'");
+        assertEvaluatedEquals("CAST(DECIMAL '12.4' AS varchar(50))", "'12.4'");
+
+        assertTrinoExceptionThrownBy(() -> evaluate("CAST(DECIMAL '12.4' AS varchar(3))"))
```

Comment:
add a test with `DECIMAL '12.40'` into `varchar(4)` (i.e. where only insginificant zero does not fit)

same for negative, and  long decimal